File: SunoFlowApp/Sources/SunoFlow/ScreenContext.swift
import Cocoa
import CoreGraphics
import Vision

/// Best-effort screen capture + on-device OCR for heuristic context.
///
/// When dictation stops, we capture the main display and run Vision's fast
/// OCR to extract the words visible on screen. This gives the cleanup LLM
/// rough context about the app the user is typing into (window titles, menu
/// items, nearby labels) so it can pick better terminology and phrasing.
///
/// Accuracy is *not* the goal — we only need the vocabulary on screen. So we
/// use Vision's `.fast` recognition level and downscale the capture to keep
/// it snappy. Everything runs on-device; only the extracted words leave the
/// machine, and they go to the cleanup endpoint alongside the transcript.
///
/// Requires the Screen Recording TCC permission. Without it,
/// `CGDisplayCreateImage` returns a valid but entirely black image (it does
/// *not* return nil), so we preflight with `CGPreflightScreenCaptureAccess`
/// and soft-fail (return nil) when permission is missing.
enum ScreenContext {
    /// Max edge length (pixels) we downscale the capture to before OCR.
    ///
    /// This was 1600px, chosen to keep a `.fast` pass cheap, and it was the
    /// single biggest cause of unusable OCR: UI text at that scale is too small
    /// to read, and the recognizer guessed. Measured on a 3420x2214 capture,
    /// only 49% of its 4+ letter words were real words, with 21 tokens carrying
    /// a digit or symbol substituted mid-word ("pa1r", "fr&sh", "invalldArant")
    /// and 16 run together with their neighbours ("restartfromdlsk").
    ///
    /// At 2400px those collapse to 0 and 2 — for the same time or slightly less,
    /// because a clearer image is less work to recognize, not more. Going beyond
    /// 2400 (measured to native 3400) buys nothing further.
    private static let maxCaptureEdge: CGFloat = 2400

    /// True if Screen Recording permission has already been granted.
    static var hasPermission: Bool { CGPreflightScreenCaptureAccess() }

    /// Triggers the system Screen Recording permission request (which opens
    /// System Settings on macOS 13+ and registers the app in the TCC list).
    /// Returns true if permission is already granted. Safe to call off the
    /// main thread.
    @discardableResult
    static func requestPermissionIfNeeded() -> Bool {
        if CGPreflightScreenCaptureAccess() { return true }
        return CGRequestScreenCaptureAccess()
    }

    /// Convenience for the Settings UI: registers the app with TCC (so it
    /// appears in the Screen Recording list) and opens the System Settings
    /// pane so the user can toggle it.
    static func openSystemSettings() {
        // Register the app with TCC first — this is what makes SunoFlow appear
        // in the Screen Recording list. Run off the main thread because the
        // call can block while the system handles the request.
        DispatchQueue.global(qos: .userInitiated).async {
            _ = CGRequestScreenCaptureAccess()
        }
        if let url = URL(string: "x-apple.systempreferences:com.apple.preference.security?Privacy_ScreenCapture") {
            NSWorkspace.shared.open(url)
        }
    }

    /// Captures the main display, runs fast on-device OCR, and returns the
    /// recognized words joined by spaces. Best-effort: returns nil on any
    /// failure (no permission, capture failed, OCR failed, no text found).
    /// Completion is called on the main queue.
    ///
    /// The screenshot and the downscale run on the background queue alongside
    /// the OCR, not on the caller's thread. This is called the instant
    /// dictation starts, when the main thread's next job is putting the
    /// recording overlay on screen, and a full Retina capture is not free.
    static func captureAndRecognize(completion: @escaping (String?) -> Void) {
        guard hasPermission else {
            completion(nil)
            return
        }
        let displayID = CGMainDisplayID()
        DispatchQueue.global(qos: .userInitiated).async {
            guard let fullImage = CGDisplayCreateImage(displayID) else {
                DispatchQueue.main.async { completion(nil) }
                return
            }
            let words = recognizeText(in: downscale(fullImage, maxEdge: maxCaptureEdge))
            DispatchQueue.main.async {
                completion(words)
            }
        }
    }

    // MARK: - OCR

    /// Runs Vision `.fast` OCR on a CGImage and returns the recognized words
    /// joined by spaces, or nil if nothing was recognized.
    private static func recognizeText(in image: CGImage) -> String? {
        let request = VNRecognizeTextRequest()
        // `.accurate`, not `.fast`. The prompt licenses the model to re-spell a
        // transcript word to match a name or identifier it finds in SCREEN, so
        // a garbled capture is not merely unhelpful — it is a source of wrong
        // corrections. Accuracy is what this input is *for*.
        //
        // It costs ~0.45s more than `.fast` (0.70s against 0.26s on a 3420x2214
        // capture), which used to be unaffordable when OCR ran between the
        // user's last word and their pasted text. It no longer does: the capture
        // starts when recording starts and finishes while they are still
        // speaking, so the cost lands in time that was being spent anyway.
        request.recognitionLevel = .accurate
        request.recognitionLanguages = ["en-US"]

        let handler = VNImageRequestHandler(cgImage: image, options: [:])
        do {
            try handler.perform([request])
        } catch {
            AppLog.log("Screen OCR failed: \(error)")
            return nil
        }
        guard let observations = request.results, !observations.isEmpty else {
            return nil
        }
        let words = observations.compactMap { $0.topCandidates(1).first?.string }
        let joined = words.joined(separator: " ")
        return joined.isEmpty ? nil : joined
    }

    // MARK: - Downscale

    /// Returns a scaled copy of `image` that fits within `maxEdge` while
    /// preserving aspect ratio. If the image is already smaller, returns it
    /// unchanged. Keeps OCR fast on large Retina captures.
    private static func downscale(_ image: CGImage, maxEdge: CGFloat) -> CGImage {
        let w = CGFloat(image.width)
        let h = CGFloat(image.height)
        let longest = max(w, h)
        guard longest > maxEdge else { return image }
        let scale = maxEdge / longest
        let newW = Int((w * scale).rounded())
        let newH = Int((h * scale).rounded())

        let colorSpace = CGColorSpaceCreateDeviceRGB()
        guard let ctx = CGContext(
            data: nil, width: newW, height: newH,
            bitsPerComponent: 8, bytesPerRow: 0,
            space: colorSpace,
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
        ) else { return image }
        ctx.interpolationQuality = .high
        ctx.draw(image, in: CGRect(x: 0, y: 0, width: newW, height: newH))
        return ctx.makeImage() ?? image
    }
}

File: apps/windows/SunoFlow/ScreenContext.cs
using System;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Imaging;
using System.Linq;
using System.Runtime.InteropServices;
using System.Runtime.InteropServices.WindowsRuntime;
using System.Threading.Tasks;
using Windows.Graphics.Imaging;
using Windows.Media.Ocr;

namespace SunoFlow;

/// <summary>
/// Screen-context OCR for the Windows tray app — the counterpart of the macOS
/// <c>ScreenContext.swift</c>. When dictation stops, captures the primary
/// display, runs on-device OCR via <c>Windows.Media.Ocr</c> (WinRT, in-box — no
/// native dep), and returns the recognized words joined by spaces. Those words
/// go to the cleanup gateway as the <c>screen</c> field so the cleanup LLM knows
/// what app/field the user is typing into → better terminology/phrasing.
///
/// <b>Accuracy is NOT the goal</b> — only the on-screen vocabulary is extracted.
/// Only the joined words leave the machine.
///
/// <b>Best-effort:</b> every failure path returns <c>""</c>. A broken OCR must
/// never break dictation. This is the hard contract, identical to the macOS
/// version's soft-fail posture.
///
/// <b>No permission layer</b> — unlike macOS (where <c>CGDisplayCreateImage</c>
/// returns a black image without the Screen Recording TCC grant), a Windows
/// foreground process can copy the screen pixels via GDI with no prompt. So the
/// <c>hasPermission</c>/<c>openSystemSettings</c> layer from the Swift version
/// is intentionally not ported. The one Windows-specific soft-fail cause is
/// "no OCR language pack installed" (<c>OcrEngine</c> is null → return <c>""</c>).
/// </summary>
internal static class ScreenContext
{
    /// Max edge (px) we downscale the capture to before OCR. Matches the macOS
    /// <c>maxCaptureEdge</c>.
    ///
    /// This was 1600, picked to keep the OCR pass cheap when it ran between the
    /// user's last word and their pasted text. It no longer runs there — the
    /// capture starts with the recording (see TrayApp.StartScreenCapture) — so
    /// the reason to keep the image small is gone, and 1600px was costing real
    /// accuracy: on a 3420x2214 macOS capture, shrinking that far left UI text
    /// too small to read and only 49% of its 4+ letter words were real words,
    /// against 76-79% at 2400.
    ///
    /// That measurement is Apple's Vision engine, not Windows.Media.Ocr, so
    /// treat the size of the win here as unverified — but the cause (text
    /// rendered too small to recognise) is not engine-specific, and the extra
    /// pixels are free now that nobody is waiting on them.
    private const int MaxCaptureEdge = 2400;

    /// <summary>
    /// Captures the primary display, runs on-device OCR, and returns the
    /// recognized words joined by spaces. Best-effort: returns <c>""</c> on any
    /// failure (capture failed, no OCR language pack, OCR threw, no text). Runs
    /// on a background thread — caller must NOT invoke on the UI thread.
    /// </summary>
    public static async Task<string> CaptureAndRecognizeAsync()
    {
        try
        {
            // 1. Capture the primary display via GDI — no permission preflight
            //    needed on Windows (direct analogue of CGDisplayCreateImage).
            // PrimaryScreen is nullable (a session with no attached display), and
            // this whole path is best-effort, so bow out rather than throw.
            var bounds = Screen.PrimaryScreen?.Bounds ?? Rectangle.Empty;
            if (bounds.Width <= 0 || bounds.Height <= 0) return "";
            using var bmp = new Bitmap(bounds.Width, bounds.Height, PixelFormat.Format32bppArgb);
            using (var g = Graphics.FromImage(bmp))
                g.CopyFromScreen(bounds.Location, Point.Empty, bounds.Size);

            // 2. Downscale to MaxCaptureEdge on the longest edge, preserving
            //    aspect ratio. Skipped if already smaller (mirrors the macOS
            //    downscale step).
            Bitmap ocrBmp = bmp;
            bool ownsOcrBmp = false;
            try
            {
                int longest = Math.Max(bmp.Width, bmp.Height);
                if (longest > MaxCaptureEdge)
                {
                    double scale = (double)MaxCaptureEdge / longest;
                    int newW = (int)Math.Round(bmp.Width * scale);
                    int newH = (int)Math.Round(bmp.Height * scale);
                    var scaled = new Bitmap(newW, newH, PixelFormat.Format32bppArgb);
                    using (var g = Graphics.FromImage(scaled))
                    {
                        g.InterpolationMode = InterpolationMode.HighQualityBicubic;
                        g.DrawImage(bmp, 0, 0, newW, newH);
                    }
                    ocrBmp = scaled;
                    ownsOcrBmp = true;
                }

                // 3. Bitmap → SoftwareBitmap via a direct BGRA buffer copy.
                //    System.Drawing's Format32bppArgb stores B-G-R-A in memory,
                //    which matches WinRT's BitmapPixelFormat.Bgra8. Stride may
                //    exceed width*4 (GDI row padding); copy row-by-row if so.
                var data = ocrBmp.LockBits(
                    new Rectangle(0, 0, ocrBmp.Width, ocrBmp.Height),
                    ImageLockMode.ReadOnly, PixelFormat.Format32bppArgb);
                byte[] pixels;
                try
                {
                    int rowBytes = data.Width * 4;
                    if (data.Stride == rowBytes)
                    {
                        pixels = new byte[rowBytes * data.Height];
                        Marshal.Copy(data.Scan0, pixels, 0, pixels.Length);
                    }
                    else
                    {
                        pixels = new byte[rowBytes * data.Height];
                        for (int y = 0; y < data.Height; y++)
                        {
                            Marshal.Copy(
                                data.Scan0 + y * data.Stride,
                                pixels, y * rowBytes, rowBytes);
                        }
                    }
                }
                finally
                {
                    ocrBmp.UnlockBits(data);
                }

                // BGRA8 + premultiplied alpha matches GDI's 32bppArgb layout.
                using var softwareBitmap = SoftwareBitmap.CreateCopyFromBuffer(
                    pixels.AsBuffer(), BitmapPixelFormat.Bgra8,
                    ocrBmp.Width, ocrBmp.Height, BitmapAlphaMode.Premultiplied);

                // 4. OCR — prefer the user's profile language, fall back to the
                //    first installed recognizer language. Null engine (no OCR
                //    language pack) → soft-fail to "".
                OcrEngine? engine = OcrEngine.TryCreateFromUserProfileLanguages();
                if (engine is null && OcrEngine.AvailableRecognizerLanguages.Count > 0)
                    engine = OcrEngine.TryCreateFromLanguage(OcrEngine.AvailableRecognizerLanguages[0]);
                if (engine is null)
                {
                    AppLog.Log("Screen OCR skipped — no OCR language pack installed");
                    return "";
                }

                var result = await engine.RecognizeAsync(softwareBitmap);

                // 5. Flatten lines → words → joined string (matches the macOS
                //    observations.compactMap{}.joined(" ")).
                if (result is null || result.Lines is null || result.Lines.Count == 0)
                    return "";
                var words = result.Lines
                    .SelectMany(l => l.Words)
                    .Where(w => !string.IsNullOrWhiteSpace(w.Text))
                    .Select(w => w.Text.Trim());
                var joined = string.Join(" ", words);
                return string.IsNullOrWhiteSpace(joined) ? "" : joined;
            }
            finally
            {
                if (ownsOcrBmp) ocrBmp.Dispose();
            }
        }
        catch (Exception ex)
        {
            AppLog.Log($"Screen OCR failed: {ex.Message}");
            return "";
        }
    }
}